plugins/languages/rust/rustaceanvim/default.nix

```
@@ -0,0 +1,62 @@
+{
+  lib,
+  helpers,
+  config,
+  pkgs,
+  ...
+}:
+with lib;
+helpers.neovim-plugin.mkNeovimPlugin config {
+  name = "rustaceanvim";
+  defaultPackage = pkgs.vimPlugins.rustaceanvim;
+
+  maintainers = [ maintainers.GaetanLepage ];
+
+  # TODO: introduced 2024-05-17, remove on 2024-02-17
+  deprecateExtraOptions = true;
+  optionsRenamedToSettings = import ./renamed-options.nix;
+
+  extraOptions = {
+    rustAnalyzerPackage = helpers.mkPackageOption {
+      name = "rust-analyzer";
+      default = pkgs.rust-analyzer;
+    };
+  };
+
+  settingsOptions = import ./settings-options.nix { inherit lib helpers pkgs; };
+
+  settingsExample = {
+    server = {
+      standalone = false;
+      cmd = [
+        "rustup"
+        "run"
+        "nightly"
+        "rust-analyzer"
+      ];
+      settings = {
+        rust-analyzer = {
+          inlayHints = {
+            lifetimeElisionHints = {
+              enable = "always";
+            };
+          };
+          check = {
+            command = "clippy";
+          };
+        };
+      };
+    };
+  };
+
+  callSetup = false;
+  extraConfig = cfg: {
+    extraPackages = [ cfg.rustAnalyzerPackage ];
+
+    plugins.rustaceanvim.settings.server.on_attach = mkDefault "__lspOnAttach";
+
+    extraConfigLua = ''
```

Would this resolve https://github.com/nix-community/nixvim/pull/1538#issuecomment-2120256132 ?

I think this is essentially what we did before.

```suggestion
    plugins.lsp.postConfig = ''
```

---

I think so

---

What if the user has not enabled `plugins.lsp` .

---

We should at least do an `assertion`/`warning` to ensure that the user has enabled `plugins.lsp`.
However, is that legit to _force_ the user to use `lspconfig.nvim` when using rustaceanvim ?

---

[upstream](https://github.com/mrcjkb/rustaceanvim/) states "No dependency on lspconfig" so I don't _think_ we need anything more than a warning? Not sure if it really makes sense to use it without LSP enabled though.

---

Ah of course. In that case we should have an assertion.

If we want to support using it without `plugins.lsp.enable` then we could do things more conditionally, e.g.:

```nix
(
  setAttrByPath
  (
    if config.plugins.lsp.enable
    then [ "plugins" "lsp" "postConfig" ]
    else [ "extraConfigLua" ]
  )
  ''vim.g.rustaceanvim = ${helpers.toLuaObject cfg.settings}''
);
```

---

Well I don't know...
"Ideally", we could just make sure that this snippet of lua is after the **eventual** LSP snippet.
We are only trying to achieve this here.
It seems overkilled to me to enforce enabling `plugins.lsp` just to make sure that the rustacenvim's config is after (and not before) the LSP block.

---

> Or... `plugins.lsp.enable = mkDefault true` in `extraConfig`. But I doubt you guys would like this.

nixpkgs does those kinda "sane defaults" quite a bit, they might even use `mkForce` in this scenario... But I think for the most part we've decided to avoid "magic" in favor of users setting things explicitly?

---

> It seems overkilled to me to enforce enabling `plugins.lsp` just to make sure that the rustacenvim's config is after (and not before) the LSP block.

That's fair.

> "Ideally", we could just make sure that this snippet of lua is after the **eventual** LSP snippet.

I don't know how we could achieve this without some complexity.

Maybe something similar to my [earlier suggestion](https://github.com/nix-community/nixvim/pull/1538#discussion_r1607235196) but also only set `on_attach` when lsp is enabled?

If you don't like using `setAttrByPath` we could just `mkMerge` multiple `mkIf` definitions instead.

---

Well if plugins.lsp is not enabled then __lspOnAttach would not exist, so we could but rustacean.nvim anywhere in the configuration. I would rather go on this path, i.e:
- If plugins.lsp.enable then the onAttach default is __lspOnAttach, and we put the setup in plugins.lsp.postConfig
- If !plugins.lsp.enable then there is no onAttach default, and we put the setup in extraConfigLua

---

This seems totally reasonable to me.